### PR TITLE
ucx: 1.13.1 -> 1.14.0 and optional ROCm support

### PR DESCRIPTION
**Pull request by bzizou**
###### Description of changes

Upgrade of ucx and added optional support for ROCm, which is necessary to compute on AMD GPUS with openmpi

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/ucx/default.nix [hunk 1]
```diff
@@ -1,7 +1,8 @@
-{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen
-, numactl, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin
+{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen, numactl
+, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin, pkgconfig
```
Comment (GabrielDougherty) [inline pkgs/development/libraries/ucx/default.nix:2]:
```suggestion
, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin, pkg-config
```

I got this error:

```
error: Function called without required argument "pkgconfig" at /home/gabriel/.cache/nixpkgs-review/pr-221347/nixpkgs/pkgs/development/libraries/ucx/default.nix:2, did you mean "pkg-config" or "pkgconf"?
(use '--show-trace' to show detailed location information)
nix --experimental-features nix-command --system x86_64-linux eval --json --impure --no-allow-import-from-derivation --expr (import /nix/store/f6qs6s51w6prb5lpz4klk85rywn4p1jb-nixpkgs-review-2.8.0/lib/python3.10/site-packages/nixpkgs_review/nix/evalAttrs.nix { allowAliases = false; attr-json = /tmp/nix-shell.tbsz32/tmpvbui2i38; }) failed to run, /tmp/nix-shell.tbsz32/tmpvbui2i38 was stored inspection
```

`pkgconfig` is a Python module. Did you perhaps mean to import `pkg-config`?

File: pkgs/development/libraries/ucx/default.nix [hunk 2]
```diff
@@ -10,29 +11,24 @@ let
...
-    #   https://github.com/openucx/ucx/pull/8450
-    (fetchpatch {
-      name = "binutils-2.39.patch";
-      url = "https://github.com/openucx/ucx/commit/6b6128efd416831cec3a1820f7d1c8e648b79448.patch";
-      sha256 = "sha256-ci00nZG8iOUEFXbmgr/5XkIfiw4eAAdG1wcEYjQSiT8=";
-    })
-  ];
-
-  nativeBuildInputs = [ autoreconfHook doxygen ];
+  nativeBuildInputs = [ autoreconfHook doxygen pkgconfig ];
```
Comment (GabrielDougherty) [inline pkgs/development/libraries/ucx/default.nix:31]:
```suggestion
  nativeBuildInputs = [ autoreconfHook doxygen pkg-config ];
```

**Review by GabrielDougherty — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/ucx/default.nix (hunk 1 above)
Comment (bzizou) [reply pkgs/development/libraries/ucx/default.nix:2]: Yes, you're right!

File: pkgs/development/libraries/ucx/default.nix [hunk 3]
```diff
@@ -1,7 +1,8 @@
-{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen
-, numactl, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin
+{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen, numactl
+, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin, pkgconfig
+, cudatoolkit, rocm-core, rocm-runtime, rocm-device-libs, hip
 , enableCuda ? false
```
Comment (markuskowa) [inline pkgs/development/libraries/ucx/default.nix:4]: The optional switches should be grouped with their dependencies like this:
```suggestion
, enableCuda ? false
, cudatoolkit
, enableRocm ? false
, rocm-core, rocm-runtime, rocm-device-libs, hip
```

**Comment by markuskowa** [comment]
Release notes: https://github.com/openucx/ucx/releases

File: pkgs/development/libraries/ucx/default.nix (hunk 1 above)
Comment (GabrielDougherty) [reply pkgs/development/libraries/ucx/default.nix:2]: this still needs to be fixed

File: pkgs/development/libraries/ucx/default.nix [hunk 4]
```diff
@@ -1,7 +1,10 @@
-{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen
-, numactl, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin
+{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen, numactl
+, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin, pkgconfig
 , enableCuda ? false
 , cudatoolkit
+, enableRocm ? false
+, rocm-core, rocm-runtime, rocm-device-libs, hip
+, enableRocm ? false
```
[the remark this answers is not in the document]
Comment (bzizou) [reply pkgs/development/libraries/ucx/default.nix]: Sorry, I just forced-pushed to fix that

File: pkgs/development/libraries/ucx/default.nix [hunk 5]
```diff
@@ -1,7 +1,9 @@
-{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen
-, numactl, rdma-core, libbfd, libiberty, perl, zlib, symlinkJoin
+{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, doxygen, numactl
```
Comment (markuskowa) [inline pkgs/development/libraries/ucx/default.nix:1]: Please remove the `fetchpatch`. It is not used anymore.
```suggestion
{ lib, stdenv, fetchFromGitHub, autoreconfHook, doxygen, numactl
```

**Comment by markuskowa** [comment]
@GrahamcOfBorg build openmpi

**Review by markuskowa — APPROVED** [review]

**Pull request merged**